server/handlers/validation.go

```diff
@@ -356,158 +373,91 @@ func validateTargets(role string, roles map[string]storage.MetaUpdate, repo *tuf
-		newKeys[kid] = data.NewPublicKey(k.Algorithm(), k.Public())
 
-		if oldRoot != nil {
-			if _, ok := oldKeys[kid]; !ok {
-				// if there is any difference in keys, a key rotation may have
-				// occurred.
-				rotation = true
-			}
-		}
-	}
-	newSigned, err := newRoot.ToSigned()
-	if err != nil {
-		return err
+	if err := signed.VerifyRoot(parsedNewSigned, newRootRole.Threshold, newRootRole.Keys); err != nil {
+		return nil, validation.ErrBadRoot{Msg: err.Error()}
 	}
-	if rotation {
-		err = signed.VerifyRoot(newSigned, oldThreshold, oldKeys)
-		if err != nil {
-			return fmt.Errorf("rotation detected and new root was not signed with at least %d old keys", oldThreshold)
-		}
-	}
-	err = signed.VerifyRoot(newSigned, newRootRole.Threshold, newKeys)
+
+	return parsedNewRoot, nil
+}
+
+// checkAgainstOldRoot errors if an invalid root rotation has taken place
+func checkAgainstOldRoot(oldRoot []byte, newRootRole data.BaseRole, newSigned *data.Signed) error {
+	parsedOldRoot := &data.SignedRoot{}
+	err := json.Unmarshal(oldRoot, parsedOldRoot)
 	if err != nil {
+		logrus.Warn("Old root could not be parsed, and cannot be used to check the new root.")
 		return err
 	}
-	root, err := data.RootFromSigned(newSigned)
+
+	oldRootRole, err := parsedOldRoot.BuildBaseRole(data.CanonicalRootRole)
 	if err != nil {
+		logrus.Warn("Old root does not have a valid root role, and cannot be used to check the new root.")
 		return err
 	}
 
-	var timestampKeyIDs []string
-
-	// at a minimum, check the 4 required roles are present
-	for _, r := range []string{rootRole, targetsRole, snapshotRole, timestampRole} {
-		role, ok := root.Signed.Roles[r]
-		if !ok {
-			return fmt.Errorf("missing required %s role from root", r)
-		}
-		// According to the TUF spec, any role may have more than one signing
-		// key and require a threshold signature.  However, notary-server
-		// creates the timestamp, and there is only ever one, so a threshold
-		// greater than one would just always fail validation
-		if (r == timestampRole && role.Threshold != 1) || role.Threshold < 1 {
-			return fmt.Errorf("%s role has invalid threshold", r)
-		}
-		if len(role.KeyIDs) < role.Threshold {
-			return fmt.Errorf("%s role has insufficient number of keys", r)
-		}
-
-		if r == timestampRole {
-			timestampKeyIDs = role.KeyIDs
+	// if the set of keys has changed between the old root and new root, then a root
+	// rotation may have taken place
+	rotation := len(oldRootRole.Keys) != len(newRootRole.Keys)
+	if !rotation { // if the number of keys is the same, we need to check every key
+		for kid := range newRootRole.Keys {
+			if _, ok := oldRootRole.Keys[kid]; !ok {
+				// if there is any difference in keys, a key rotation may have
+				// occurred.
+				rotation = true
+			}
 		}
 	}
 
-	// ensure that at least one of the timestamp keys specified in the role
-	// actually exists
-
-	for _, keyID := range timestampKeyIDs {
-		if timestampKey.ID() == keyID {
-			return nil
+	if rotation {
```

Comment:
We should always verify against old regardless of whether we detect a rotation or not. Spec requires that any new root verifies against the previous one.